pkg/codegen/report/report.go

```diff
@@ -29,9 +29,11 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
 	hcl2 "github.com/pulumi/pulumi/pkg/v3/codegen/pcl"
 	"github.com/pulumi/pulumi/pkg/v3/version"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/env"
 )
 
-const ExportTargetDir = "PULUMI_CODEGEN_REPORT_DIR"
+var ExportTargetDir = env.String("CODEGEN_REPORT_DIR",
```

Comment:
For the purpose of hygiene and testability, I'd recommend moving environment variables into an `Environment` struct, and having one global instance of the environment struct rather than many global `Variable` structs. 

Then, it's optional whether the `Environment` object could be initialized in an `init` function or at the top of `main`. 

The primary benefit IMO is that you can inject the `Environment` object into functions that require one or more env vars; this lets you test control flow that uses environment variables without having to test them in your tests (which can be error-prone, since it makes the test impure).

Reply:
> For the purpose of hygiene and testability, I'd recommend moving environment variables into an Environment struct, and having one global instance of the environment struct rather than many global Variable structs.

I'm not sure that I see the benefit of a global instance over module level variables here. It also forces each variable to be declared in the same scope, vs where they are used.

> Then, it's optional whether the Environment object could be initialized in an init function or at the top of main.

In general, I think we should avoid tests depending on env vars. If functions need to be explicitly set to up to take the env bundle, they should take just the values they need as arguments. If we want to, we can add support for creating constant values that conform to `StringValue` and `BoolValue`. These can be assigned before testing.

> The primary benefit IMO is that you can inject the Environment object into functions that require one or more env vars; this lets you test control flow that uses environment variables without having to test them in your tests (which can be error-prone, since it makes the test impure).

Can you clarify. I don't think that go supports the kind of injection I'm thinking of here.

Reply:
Ok, here's my pitch to explain the benefit. It mirrors the definition of the go HTTP server; there's a global instance you can summon when you need it, and ignore when you don't.

Here's the benefit.

Suppose I wanted write a function that checks if pulumi-experimental is enabled and does something.

```golang
func skip_checkpoints() {
   if getModuleLevelEnvVar() == "1"  {
       // ...do the thing
   }
}
```

This requires global state, which makes the test unable to be run in parallel, since it could conflict with other tests also modifying that global state in the same process. 

Instead, you could write:
```golang
func skipCheckpoints(env Environment) {
   if env.Get("PULUMI_EXPERIMENTAL") == "1"  {
       // ...do the thing
   }
}
```
and perform dependency injection of the Environment. Then you'd wrap your original code like this:

```golang
func skipCheckpointsWrapper() {
    var env = getGlobalEnv()
    skipCheckpoints(env)
}
```

The wrapper is what you'd use in your client code, but the version with D.I. is what you'd cover with a unit test. In general, we want unit tests to be pure and total so they can be isolated, so minimizing global state improves testability. That's the general goal of D.I. in this context.

--------

> It also forces each variable to be declared in the same scope, vs where they are used.

Yes. I believe all of our environment variables should be declared in a single location. I don't think we want env variable declarations scattered throughout the code. We should declare all configurables upfront in a single location where possible. This improves discoverability. It's true that means they are "further away" from where they're used, but their location is standardized, and only a single hop ("Jump to Definition") that colocates all of the environment variables. 

> In general, I think we should avoid tests depending on env vars. If functions need to be explicitly set to up to take the env bundle, they should take just the values they need as arguments

I agree that it's true we want to minimize unit tests that exercise features with environment variables. In practice we already have a number of locations that use them. To see those locations, enable this linter in golangci-lint: https://github.com/sivchari/tenv . Those are just the places that set env vars in tests that violate the lints.

As you move up the test pyramid, you'll inevitably have more coarse-grained tests that exercise larger sections of the codebase, some of which will test functionality with configurable toggles. The way to keep this most pure is to inject some `config` struct at each of those sites: i.e. each subsystem of Pulumi is initialized with the set of configurables it needs, following RAII and "fail fast" design methodologies. I agree that if you do that, you won't need to ever set env vars in your tests, because you can always inject your own value. But I worry we aren't far enough along to make that happen, as evidenced by the failing lints.

Does that make sense? I guess I'm trying to offer feedback on two fronts:
1. Why grouping together env vars in a single location is better than spreading them out.
2. Why having a single global Environment instance is better than having a collection of env vars instances (D.I. becomes more straightforward for higher-up-the-pyramid tests).

Reply:
> Yes. I believe all of our environment variables should be declared in a single location. I don't think we want env variable declarations scattered throughout the code. We should declare all configurables upfront in a single location where possible. This improves discoverability. It's true that means they are "further away" from where they're used, but their location is standardized, and only a single hop ("Jump to Definition") that colocates all of the environment variables.

While I get some of Ian's hesitation towards this, I think this would be more inline with how production ready configuration libraries generally work. 

It is nice being able to define the env vars at the site of use, but 
A) We should probably be adding some friction to adding new envvars, they aren't free and every axis of configuration is an axis to worry about testing and maintaining
B) Being able to inject the environment as a configuration object of all config rather than ad-hoc access to getenv would be a _massive_ unit testing win
C) Up front validation of configuration would be nice. I think there's some debate here about if for example an envvar is set invalid but then doesn't get used wouldn't it be nicer to ignore the error rather than telling the user to fix it, but I'm not sure that's an overall win.

Reply:
A) I think that this friction is the job of code review. I don't think it needs to be technologically actually difficult. I have centralized env var declaration to a single file.

B) We can do this. I have included an example in a new commit.

C) We can do this. To demonstrate functionality, I have added an experimental `pulumi env` command, which returns a failing exit code when a variable has an invalid value. It warns for suspicious values, like booleans with non-standard false values. 


```
$ PULUMI_DEV=3 pulumi env
warning: PULUMI_DEV: "3" is falsy, but doesn't look like a boolean
Variable                       Description                                         Value
PULUMI_DEV                     Enable features for hacking on pulumi itself        false
PULUMI_EXPERIMENTAL            Enable experimental options and commands            unset
PULUMI_IGNORE_AMBIENT_PLUGINS  Discover additional plugins by examining the $PATH  unset
PULUMI_SKIP_UPDATE_CHECK       Disable checking for a new version of pulumi        unset